src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -267,34 +306,170 @@ private ValueTask<(HttpConnection, HttpResponseMessage)> GetConnectionAsync(Http
             }
         }
 
+        private async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+            GetHttp2ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_kind == HttpConnectionKind.Https || _kind == HttpConnectionKind.SslProxyTunnel);
+
+            // See if we have an HTTP2 connection
+            Http2Connection http2Connection = _http2Connection;
+            if (http2Connection != null)
+            {
+                if (NetEventSource.IsEnabled) Trace("Using existing HTTP2 connection.");
+                return (http2Connection, false, null);
+            }
+
+            // Ensure that the connection creation semaphore is created 
+            if (_http2ConnectionCreateLock == null)
+            {
+                lock (SyncObj)
+                {
+                    if (_http2ConnectionCreateLock == null)
+                    {
+                        _http2ConnectionCreateLock = new SemaphoreSlim(1);
+                    }
+                }
+            }
+
+            // Try to establish an HTTP2 connection
+            Socket socket = null;
+            SslStream sslStream = null;
+            TransportContext transportContext = null;
+
+            // Serialize creation attempt
+            await _http2ConnectionCreateLock.WaitAsync().ConfigureAwait(false);
+            try
+            {
+                if (_http2Connection != null)
+                {
+                    // Someone beat us to it
+
+                    if (NetEventSource.IsEnabled)
+                    {
+                        Trace("Using existing HTTP2 connection.");
+                    }
+
+                    return (_http2Connection, false, null);
+                }
+
+                // Recheck if HTTP2 has been disabled by a previous attempt.
+                if (_http2Enabled)
+                {
+                    if (NetEventSource.IsEnabled)
+                    {
+                        Trace("Attempting new HTTP2 connection.");
+                    }
+
+                    Stream stream;
+                    HttpResponseMessage failureResponse;
+                    (socket, stream, transportContext, failureResponse) =
+                        await ConnectAsync(request, true, cancellationToken).ConfigureAwait(false);
+                    if (failureResponse != null)
+                    {
+                        return (null, true, failureResponse);
+                    }
+
+                    sslStream = (SslStream)stream;
+                    if (sslStream.NegotiatedApplicationProtocol == SslApplicationProtocol.Http2)
+                    {
+                        // The server accepted our request for HTTP2.
```

Comment:
Verify the negotiated TLS version >= 1.2.

Reply:
Re ReferenceEquals: the method is "normalized" before calling this so I believe it's fine in this case.

Re disabling 1.1, no plans currently.

Re TLS version, will do.